mac80211: ath: improve ath10k "failed to flush transmit queue" errors

maurerle opened the pull request:
Currently, in busy environments, ath10k and ath11k log "failed to flush transmit queue" errors and have a spiking CPU usage, making the wireless barely usable. With this patch, taken from https://patchwork.kernel.org/project/linux-wireless/list/?series=911851 - this does not occur in normal operation for ath10k anymore and the wifi is much more stable.

Would be very good to have this in OpenWrt.

Supersedes #19427
Ref: https://github.com/openwrt/openwrt/issues/15929#issuecomment-2227391307 Ref: https://github.com/openwrt/openwrt/issues/14965

neheb commented:
this patch looks like it's for ath10k only.

neheb commented:
Also,

```
make package/mac80211/{clean,refresh}
```

castiel652 commented:
Did you merge those two patches into one?

maurerle commented:
That happened with https://github.com/openwrt/openwrt/pull/19529#issuecomment-3121003324
I don't know if there is a way to keep the patch structure of the series

egc112 reviewed (approved):
Compiled with your patch for my R7800 with 24.10 snapshot
My R7800 runs without a problem.
It is just a test router and only tested with my phone on 5 Ghz, iperf throughput is 650 Mb/s which is excellent (AC 80 Mhz two stream, phy rate 866)
Log is clean but only a short test without any other clients and traffic so cannot say for sure if this solves the problem you mentioned but at least it is working normally in this short test
As this is a patch from upstream I am OK with it so lets go and try it :)

castiel652 commented:
Well I did some more reading and I think the second patch is redundant.
The only place that call `__ieee80211_flush_queues` with `queues` set is `ieee80211_reserve_tid` and it's not called anywhere in mac80211 or ath10k.

neheb commented:
I assume given how widespread this problem is that this should be merged soon.

neheb commented:
So with this PR as is, I'm getting

```
failed to flush sta txq
```

different error?

neheb commented:
oh @repk is on here.

hauke commented:
You merged both patches into one:
https://patchwork.kernel.org/project/linux-wireless/patch/17d26d6a3e80ff03939ee7935fdc07f979b61a4f.1732293922.git.repk@triplefau.lt/
https://patchwork.kernel.org/project/linux-wireless/patch/01d859e8e574a1f5d0b916333fe0b5cda859af9b.1732293922.git.repk@triplefau.lt/

Please keep both individual patches, no problem adding two patches in one commit.

Is this change also needed for ath10k-ct?

You said you also tested the other patches:
https://patchwork.kernel.org/project/linux-wireless/patch/20250806070005.1429-1-hujy652@gmail.com/#26508138
Is this better than the one you proposed? 

I think we should merge this, it looks like QCA does not care about ath10k any more and mostly blocks all changes.

maurerle commented:
@hauke thanks for getting back to this.
I did not test this on ath10k-ct in our environment unfortunately.
Code-quality wise the new pr looks cleaner, behaviorwise they both fix the same issue.
So i will prepare a PR for this instead.

maurerle committed:
mac80211: ath: improve ath10k "failed to flush transmit queue" errors

Currently, in busy environments, ath10k logs "failed to flush transmit
queue" errors and have a spiking CPU usage, making the wireless barely
usable.
With this patch, taken from https://patchwork.kernel.org/project/linux-wireless/patch/20250806070005.1429-1-hujy652@gmail.com/
this does not occur in normal operation and the wifi is much more stable

Supersedes #19427
Ref: https://github.com/openwrt/openwrt/issues/15929#issuecomment-2227391307
Ref: https://github.com/openwrt/openwrt/issues/14965

Signed-off-by: Florian Maurer <f.maurer@outlook.de>
Link: https://github.com/openwrt/openwrt/pull/19529
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

hauke commented:
If Qualcomm merges a different version to fix this bug upstream, please create a PR to update OpenWrt to use the updated version.

castiel652 commented:
> Is this change also needed for ath10k-ct?
@hauke ath10k-ct doesn't need this. It has its own flushing mechanism.

nazar-pc commented:
Nice, I'll test it out when new release comes out or when I have a chance to test a snapshot and will close that issue once I confirm it works well. Thanks!
